packages/vite/src/node/ssr/ssrTransform.ts

```diff
@@ -81,15 +86,13 @@ async function ssrTransformScript(
   try {
     ast = await rollupParseAstAsync(code)
   } catch (err) {
-    if (!err.loc || !err.loc.line) throw err
-    const line = err.loc.line
-    throw new Error(
-      `Parse failure: ${
-        err.message
-      }\nAt file: ${url}\nContents of line ${line}: ${
-        code.split('\n')[line - 1]
-      }`,
-    )
+    if (err.code === 'PARSE_ERROR' && typeof err.pos === 'number') {
+      err.id = url
+      err.loc = numberToPos(code, err.pos)
+      err.loc.file = url
+      err.frame = generateCodeFrame(code, err.pos)
```

Comment:
I think we should keep the error message (`Parse failure: `...) so that `console.log(err)` shows the position and contents (#5192, #12060).

Reply:
I don't know when `err.loc` exsits (when it passes the previous code path) though.

Reply:
Right, it looks better to keep `Parse failure: ...` for clarity. I remembered esbuild transform error also includes location details in its message (cf. https://github.com/vitejs/vite/pull/18626).

> I don't know when `err.loc` exsits (when it passes the previous code path) though.

Probably `loc` check was for acorn based parser error, so it's no longer relevant. Since Rollup 4, it now only has `code, message, pos` properties https://github.com/rollup/rollup/blob/42e587e0e37bc0661aa39fe7ad6f1d7fd33f825c/src/utils/bufferToAst.ts#L20-L22, so we only need to target these. It looks like `pos` is missing on rust side panic, so I covered that case too.

Reply:
Thanks for checking! That makes sense.